src/util/threadpool.rs

```diff
@@ -0,0 +1,610 @@
+    // `group_concurrency[group_id]` tasks of the group are running.
+    group_concurrency: HashMap<T, usize>,
+    // max num of threads each group can run on when pool is busy.
+    // each value in group_concurrency shouldn't bigger than this value.
+    group_concurrency_on_busy: usize,
+}
+
+impl<T: Hash + Eq + Send + Clone> BigGroupThrottledQueue<T> {
+    pub fn new(group_concurrency_on_busy: usize) -> BigGroupThrottledQueue<T> {
+        BigGroupThrottledQueue {
+            group_concurrency: HashMap::default(),
+            waiting_queue: HashMap::default(),
+            pending_tasks: BinaryHeap::new(),
+            group_concurrency_on_busy: group_concurrency_on_busy,
+        }
+    }
+
+    // Try push into pending. Return none on success,return Some(task) on failed.
+    #[inline]
+    fn try_push_into_pending(&mut self, task: Task<T>) -> Option<Task<T>> {
+        let count = self.group_concurrency.entry(task.group_id.clone()).or_insert(0);
+        if *count >= self.group_concurrency_on_busy {
+            return Some(task);
+        }
+        *count += 1;
+        self.pending_tasks.push(task);
+        None
+    }
+
+    #[inline]
+    fn pop_task_from_waiting_queue(&mut self) -> Option<Task<T>> {
+        let group_id = self.pop_group_id_from_waiting_queue();
+        if group_id.is_none() {
+            return None;
+        }
+        let group_id = group_id.unwrap();
+        let task = self.pop_from_waiting_queue_with_group_id(&group_id);
+        // update group_concurrency since current task is going to run.
+        let mut count = self.group_concurrency.entry(group_id.clone()).or_insert(0 as usize);
+        *count += 1;
+        Some(task)
+    }
+
+    #[inline]
+    fn pop_from_waiting_queue_with_group_id(&mut self, group_id: &T) -> Task<T> {
+        let (left_tasks_num, task) = {
+            let mut waiting_tasks = self.waiting_queue.get_mut(group_id).unwrap();
+            let task = waiting_tasks.pop_front().unwrap();
+            (waiting_tasks.len(), task)
+        };
+        // if waiting tasks for group is empty,remove from waiting_tasks.
+        if left_tasks_num == 0 {
+            self.waiting_queue.remove(group_id);
+        }
+        task
+    }
+
+    // pop_group_id_from_waiting_queue returns the next task's group_id.
+    // we choose the group according to the following rules:
+    // 1. Choose the groups with least running tasks.
+    // 2. If more than one groups has the least running tasks,
+    //    choose the one whose task comes first(with the minum task's ID)
+    #[inline]
+    fn pop_group_id_from_waiting_queue(&mut self) -> Option<T> {
+        // (group_id,count,task_id)
+        let mut next_group = None;
+        for (group_id, tasks) in &self.waiting_queue {
+            let front_task_id = tasks[0].id;
+            assert!(self.group_concurrency.contains_key(group_id));
+            let count = self.group_concurrency[group_id];
```

Comment:
Yes. The task should be pushed into `pending_tasks`  if `group_id` not in `self.group_concurrency` or `self.group_concurrency[group_id]<group_concurrency_on_busy`. So we can ensure the group_id exist here @siddontang